source/common/crypto/utility_impl.cc

```diff
@@ -35,6 +35,86 @@ std::vector<uint8_t> UtilityImpl::getSha256Hmac(const std::vector<uint8_t>& key,
   return hmac;
 }
 
+const EncryptionDecryptionOutput UtilityImpl::decrypt(CryptoObject& key,
+                                                      const std::vector<uint8_t>& cipher_text) {
+  // Step 1: get private key
+  auto pkey_wrapper = Common::Crypto::Access::getTyped<Common::Crypto::PrivateKeyObject>(key);
+  EVP_PKEY* pkey = pkey_wrapper->getEVP_PKEY();
+  if (pkey == nullptr) {
+    return {false, "failed to get private key for decryption"};
+  }
+
+  // Step 2: initialize EVP_PKEY_CTX
+  EVP_PKEY_CTX* ctx = EVP_PKEY_CTX_new(pkey, nullptr);
```

Comment:
May be we can create a simple struct which will call the `EVP_PKEY_CTX_free(ctx)` automatically when it's destructed.
Then we needn't call `EVP_PKEY_CTX_free(ctx)` everywhere.